server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1310,18 +1413,31 @@ private static Map<String, IndexAbstraction> filterIndicesLookupForSuffixWildcar
 
         /**
          * Return the {@code Stream} of open and/or closed index names for the given {@param resources}.
-         * Datastreams and aliases are interpreted to refer to multiple indices,
+         * Data streams and aliases are interpreted to refer to multiple indices,
          * then all index resources are filtered by their open/closed status.
          */
         private static Stream<String> expandToOpenClosed(Context context, Stream<IndexAbstraction> resources) {
             final IndexMetadata.State excludeState = excludeState(context.getOptions());
             return resources.flatMap(indexAbstraction -> {
                 if (context.isPreserveAliases() && indexAbstraction.getType() == Type.ALIAS) {
+                    // TODO what if it's a data stream alias
                     return Stream.of(indexAbstraction.getName());
                 } else if (context.isPreserveDataStreams() && indexAbstraction.getType() == Type.DATA_STREAM) {
+                    // TODO if we only want failure store should we only return data streams that have failure store?
```

Comment:
Looking at the code it seems that if we are setting preserve data streams then it is some kind of operation that is only running against data streams (data stream stats, creating aliases, snapshotting, etc). In those cases I wonder if it doesn't entirely make sense to allow for setting the failure store flag? We may need to approach those API's on a case by case basis.